Sources/SwiftBSON/BSONEncoder.swift

```diff
@@ -859,46 +841,20 @@ private class MutableDictionary: BSONValue {
     }
 
     /// Converts self to a `BSONDocument` with equivalent key-value pairs.
-    fileprivate func toDocument() -> BSONDocument {
+    fileprivate func toDocument() throws -> BSONDocument {
         var doc = BSONDocument()
         for i in 0..<self.keys.count {
             let value = self.values[i]
-            doc.append(key: self.keys[i], value: value.bson)
+            doc.append(key: self.keys[i], value: try value.toBSON())
```

Comment:
`doc.append` will start throwing as part of the null bytes fix, because it calls through to `ByteBuffer.writeCString` which will start throwing as well.
altogether, this will allow us to error if a user tries to encode something containing a null byte in one of the coding keys, rather than crashing.